src/fairseq2/nn/lora.py

```diff
@@ -0,0 +1,373 @@
+
+    def forward(self, x: Tensor) -> Tensor:
+        if self.merged:
+            return embedding(x, self.weight, self.pad_idx)
+        else:
+            return embedding(
+                x,
+                self.weight + (self.lora_B @ self.lora_A).T * self.scaling,
+                self.pad_idx,
+            )
+
+    def reset_lora_parameters(self) -> None:
+        """Reset the parameters and buffers of the module."""
+        nn.init.zeros_(self.lora_A)
+        nn.init.normal_(self.lora_B)
+
+    @property
+    def wrapped_module(self) -> nn.Module:
+        return self
+
+    def merge(self) -> None:
+        if self.merged:
+            return
+
+        self.weight.data += (self.lora_B.data @ self.lora_A.data).T * self.scaling
+        self.merged = True
+
+    def unmerge(self) -> None:
+        if not self.merged:
+            return
+
+        self.weight.data -= (self.lora_B.data @ self.lora_A.data).T * self.scaling
+        self.merged = False
+
+
+@final
+class LoRALinear(Projection, LoRALayer):
+    weight: Parameter
+    bias: Optional[Parameter]
+    lora_A: Parameter
+    lora_B: Parameter
+    dropout: Optional[Dropout]
+    skip_init: bool
+    merged: bool
+
+    def __init__(
+        self,
+        wrapped: Projection,
+        config: LoRAConfig,
+        skip_init: bool = False,
+        device: Optional[Device] = None,
+        dtype: Optional[DataType] = None,
+    ) -> None:
+        """
+        :param wrapped:
+            The Linear module to be wrapped.
+        :param config:
+            The LoRA config.
+        :param skip_init:
+            If ``True``, the weights and bias will be left uninitialized, and
+            :meth:`reset_lora_parameters` will become noop.
+        """
+        Projection.__init__(self, wrapped.input_dim, wrapped.output_dim)
+
+        LoRALayer.__init__(self, config)
+
+        self.weight = Parameter(
+            wrapped.weight, requires_grad=False  # type: ignore[arg-type]
+        )
+
+        if wrapped.bias is not None:
+            self.bias = Parameter(
+                wrapped.bias, requires_grad=False  # type: ignore[arg-type]
+            )
+        else:
+            self.register_module("bias", None)
+
+        self.lora_A = Parameter(
+            torch.empty((self.r, self.input_dim), device=device, dtype=dtype)
+        )
+
+        self.lora_B = Parameter(
+            torch.empty((self.output_dim, self.r), device=device, dtype=dtype)
+        )
+
+        if self.dropout_p > 0.0:
+            self.dropout = Dropout(self.dropout_p)
+
+        else:
+            self.register_module("dropout", None)
+
+        self.merged = False
+
+        self.skip_init = skip_init
+
+        self.reset_lora_parameters()
+
+    def forward(self, x: Tensor) -> Tensor:
+        if self.merged:
+            return linear(x, self.weight, self.bias)
+        else:
+            h1 = linear(x, self.weight, self.bias)
+
+            if self.dropout is not None:
```

Comment:
To avoid repetitiveness, you can say:

```python
...
if self.dropout is not None:
  x = self.dropout(x)

h2 = linear(x, self.lora_B @ self.lora_A * self.scaling)
...